### ISSUE TITLE
Generated Docker images have a single layer

### ISSUE THREAD

**Issue by nbraud** [andsens/bootstrap-vz#340]
Docker images constitute of several layers, filesystem overlays that are combined together to produce the actual image's filesystem.  The rationale for doing this is two-fold:

1. Multiple images can share layers (say, a bare `debootstrap`ed Debian install can be common to multiple images) to save disk space (and cache).
2. Layers are an integral part of how Docker lets you reuse images by slapping more customizations on top.

While 2. is not relevant (I think) for bootstrap-vz, 1. very likely is, and producing *reuseable* image layers would be great.  However, the way Docker implements this (assuming that running the same command in the same layer produces the same result) has [notorious issues](https://docs.docker.com/engine/userguide/eng-image/dockerfile_best-practices/#apt-get) and is likely not applicable here.

However, something that could be done is to run the build as usual (without Docker-style caching) and switch to a different layer at the end of a few phases, such as `os_installation`, `package_installation`, `system_modification` and `user_modification`.
Provided that the installation of the same packages (in the same version and so on...) results in the same layer (in other words, provided that the layer *builds reproducibly*), then the image can share layers with other images that use the same debootstrap parameters, install the same packages, ...

**Comment by nbraud** [comment]
@andsens This is, right now, a mildly wild idea, in the sense that I didn't even check yet if very simple images build reproducibly (and if not, where the unreproducibility lies).  However, I think reproducible builds would be a very powerful feature to have.

I filed the issues so that other people may read it (and that I do not forget).

**Comment by andsens** [comment]
That is a great idea. I love the concept of `commit`s between phases.  
We would have to leverage dockers built-in concept of caching in order to reuse layers, which is where I think it would get tricky. Alternatively we'd need some way of uniquely identifying a specific build setup ourselves, which I think would be even harder (things would slip through the cracks all the time).

Still, this is definitely worth discussing :-)

**Comment by nbraud** [comment]
> We would have to leverage dockers built-in concept of caching in order to reuse layers, which is where I think it would get tricky.

Exactly my point: Docker's caching mechanism is utterly unsafe/insane.
My proposal to deal with it would be to run the entire build (without caching) but make the phases deterministic: that way, the same layers get built, get the same id (as the id is a hash of the layer) and get only stored once.

**Comment by andsens** [comment]
> that way, the same layers get built, get the same id (as the id is a hash of the layer) and get only stored once.

You won't get the same ID, the timestamps of files change from run to run :-(

**Comment by nbraud** [comment]
@andsens Yes; as said originally, it will need work to make the phases deterministic.
I had a first look with `diffoscope`, and it looks pretty promising: the only differences when building a [small container](https://nicolas.braud-santoni.eu/tmp/mini-docker.yaml) are [relatively small](https://nicolas.braud-santoni.eu/tmp/diffoscope.log).
The differences are:

- A few files are non-deterministic, but can be thrown away:
  - `/etc/machine-id` is a random id;
  - `/etc/init.d/.depend.{boot,stop}` has non-deterministic order;
  - `/var/log/alternatives.log` has timestamps;
  - `/var/cache/ldconfig/aux-cache`.
- `/bin/{compress,gunzip}` are for some not-yet-investigated reason different.
- Many files (not coming from a package directly) have differing mtimes.
